In [1]:
import pandas as pd
import numpy as np

file_path = "../data/netflix_titles.csv"
df = pd.read_csv(file_path)

df.head()

,show_id,type,title,director,cast,country,date_added,release_year,rating,duration,listed_in,description
0,s1,Movie,Dick Johnson Is Dead,Kirsten Johnson,NaN,United States,"September 25, 2021",2020,PG-13,90 min,Documentaries,"As her father nears the end of his life, filmm..."
1,s2,TV Show,Blood & Water,NaN,"Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...",South Africa,"September 24, 2021",2021,TV-MA,2 Seasons,"International TV Shows, TV Dramas, TV Mysteries","After crossing paths at a party, a Cape Town t..."
2,s3,TV Show,Ganglands,Julien Leclercq,"Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...",NaN,"September 24, 2021",2021,TV-MA,1 Season,"Crime TV Shows, International TV Shows, TV Act...",To protect his family from a powerful drug lor...
3,s4,TV Show,Jailbirds New Orleans,NaN,NaN,NaN,"September 24, 2021",2021,TV-MA,1 Season,"Docuseries, Reality TV","Feuds, flirtations and toilet talk go down amo..."
4,s5,TV Show,Kota Factory,NaN,"Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...",India,"September 24, 2021",2021,TV-MA,2 Seasons,"International TV Shows, Romantic TV Shows, TV ...",In a city of coaching centers known to train I...


In [2]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 8807 entries, 0 to 8806
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype
---  ------        --------------  -----
 0   show_id       8807 non-null   str  
 1   type          8807 non-null   str  
 2   title         8807 non-null   str  
 3   director      6173 non-null   str  
 4   cast          7982 non-null   str  
 5   country       7976 non-null   str  
 6   date_added    8797 non-null   str  
 7   release_year  8807 non-null   int64
 8   rating        8803 non-null   str  
 9   duration      8804 non-null   str  
 10  listed_in     8807 non-null   str  
 11  description   8807 non-null   str  
dtypes: int64(1), str(11)
memory usage: 825.8 KB


Why datatype for date_added is str instead of date
Some fields have null value

In [3]:
df.describe()

,release_year
count,8807.000000
mean,2014.180198
std,8.819312
min,1925.000000
25%,2013.000000
50%,2017.000000
75%,2019.000000
max,2021.000000


In [4]:
df.describe(include="object")

/var/folders/3p/13n6clgx5432cc1mbk90nkv00000gn/T/ipykernel_18713/702825166.py:1: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  df.describe(include="object")


,show_id,type,title,director,cast,country,date_added,rating,duration,listed_in,description
count,8807,8807,8807,6173,7982,7976,8797,8803,8804,8807,8807
unique,8807,2,8807,4528,7692,748,1767,17,220,514,8775
top,s1,Movie,Dick Johnson Is Dead,Rajiv Chilaka,David Attenborough,United States,"January 1, 2020",TV-MA,1 Season,"Dramas, International Movies","Paranormal activity at a lush, abandoned prope..."
freq,1,6131,1,19,19,2818,109,3207,1793,362,4


How come there is 748 unique values for "country"? total number of countries in the world is 195! look for:
- case sensitivity
- space
- mis spelling

In [5]:
# count missing values
df.isnull().sum()

show_id            0
type               0
title              0
director        2634
cast             825
country          831
date_added        10
release_year       0
rating             4
duration           3
listed_in          0
description        0
dtype: int64

In [6]:
# Calculate missing percentage
(df.isnull().sum()/len(df))*100

show_id          0.000000
type             0.000000
title            0.000000
director        29.908028
cast             9.367549
country          9.435676
date_added       0.113546
release_year     0.000000
rating           0.045418
duration         0.034064
listed_in        0.000000
description      0.000000
dtype: float64

Top 3 fields with the most missing data: director, then cast and country.

In [7]:
# any duolicates?
df.duplicated().sum()

np.int64(0)

## Summary of the Initial Dataset Observations

- Dataset contains 8,807 rows and 12 columns.
- `date_added` is stored as text and should be converted to a datetime type.
- The `director` column has a significant number of missing values.
- No duplicate rows were found.
- The dataset includes both numerical and categorical columns.
- The earliest release year is XXXX and the latest is YYYY.

# Phase 2: Cleaning Data

In [8]:
# Make a copy of the original data
df_clean = df.copy()

In [9]:
df_clean.isnull().sum()

show_id            0
type               0
title              0
director        2634
cast             825
country          831
date_added        10
release_year       0
rating             4
duration           3
listed_in          0
description        0
dtype: int64

In [ ]:
# Let'd decide about some of these missing values
# this dataset is mostly text so replacing missing values with a default text makes sense
# if the value was numerical we could .fillna(mean) or .fillna(median) based on the logic of situation
df_clean["director"] = df_clean["director"].fillna("Unknown")

In [12]:
df_clean.dtypes

show_id           str
type              str
title             str
director          str
cast              str
country           str
date_added        str
release_year    int64
rating            str
duration          str
listed_in         str
description       str
dtype: object

In [14]:
df_clean["date_added"] = pd.to_datetime(df_clean["date_added"], format="mixed")

In the above cell at first I didn't pass the format argument, it threw an error about the format of a few records that it couldn't match. with this argument I stated that there could be inconsistancies in different rows of this value and expected it to comprehend this. 
In real project I should keep record of these records and make sure the data/time is converted accordingly. As another approach I may use another format to enforce a specific conversion.

In [15]:
df_clean.dtypes

show_id                    str
type                       str
title                      str
director                   str
cast                       str
country                    str
date_added      datetime64[us]
release_year             int64
rating                     str
duration                   str
listed_in                  str
description                str
dtype: object

## Feature engineering
Let's add year, month, and day of adding the data from the date_added

In [19]:
df_clean["year_added"] = df_clean["date_added"].dt.year
df_clean["month_added"] = df_clean["date_added"].dt.month_name()
df_clean["day_added"] = df_clean["date_added"].dt.day_name()

In [20]:
df_clean.sample(10)

,show_id,type,title,director,cast,country,date_added,release_year,rating,duration,listed_in,description,year_added,month_added,day_added
2246,s2247,TV Show,Down to Earth with Zac Efron,Unknown,"Zac Efron, Darin Olien",United States,2020-07-10,2020,TV-PG,1 Season,"Docuseries, Reality TV, Science & Nature TV",Actor Zac Efron journeys around the world with...,2020.0,July,Friday
8307,s8308,Movie,The Forbidden Kingdom,Rob Minkoff,"Jackie Chan, Jet Li, Collin Chou, Liu Yifei, L...","United States, China",2019-07-01,2008,PG-13,104 min,"Action & Adventure, Sci-Fi & Fantasy",Jackie Chan and Jet Li star in this rousing ad...,2019.0,July,Monday
3168,s3169,TV Show,Fuller House,Unknown,"Candace Cameron Bure, Jodie Sweetin, Andrea Ba...",United States,2019-12-06,2020,TV-PG,5 Seasons,TV Comedies,The Tanner family’s adventures continue as DJ ...,2019.0,December,Friday
4423,s4424,Movie,ABCD: Any Body Can Dance,Remo D'Souza,"Prabhu Deva, Kay Kay Menon, Ganesh Acharya, Dh...",India,2018-11-01,2013,TV-14,136 min,"Dramas, International Movies",Thrown out of the elite dance studio he founde...,2018.0,November,Thursday
8278,s8279,Movie,The Devil We Know,"Stephanie Soechtig, Jeremy Seifert",NaN,United States,2019-01-15,2018,TV-14,88 min,Documentaries,This documentary details DuPont's alleged deca...,2019.0,January,Tuesday
3516,s3517,Movie,Oh! Baby,B. V. Nandini Reddy,"Samantha Ruth Prabhu, Lakshmi, Rajendraprasad,...",India,2019-09-14,2019,TV-14,157 min,"Comedies, International Movies, Music & Musicals",A surly septuagenarian gets another chance at ...,2019.0,September,Saturday
5013,s5014,Movie,Something Huge,Carlo Padial,"Berto Romero, Carolina Bang, Javier Botet, Car...",Spain,2018-03-01,2017,TV-MA,87 min,"Comedies, Independent Movies, International Mo...","A director and a comedian want to shoot the ""c...",2018.0,March,Thursday
2502,s2503,Movie,Escaping Tel Aviv,Sherif Arafa,"Karim Abdel Aziz, Mona Zaki, Sherif Mounir, En...",Egypt,2020-05-19,2009,TV-14,120 min,"Action & Adventure, International Movies",After her husband reveals he's an undercover M...,2020.0,May,Tuesday
6896,s6897,Movie,Green Room,Jeremy Saulnier,"Anton Yelchin, Imogen Poots, Patrick Stewart, ...",United States,2018-11-12,2015,R,95 min,"Horror Movies, Thrillers","At the end of their low-budget tour, a punk ba...",2018.0,November,Monday
4464,s4465,Movie,Welcome to Sajjanpur,Shyam Benegal,"Shreyas Talpade, Amrita Rao, Ravi Kishan, Ila ...",India,2018-11-01,2008,TV-14,131 min,"Comedies, Dramas, International Movies",An educated man assists the illiterate residen...,2018.0,November,Thursday
